DRAFT: Eschan_Portal_#1 (Escha-Ru'Aun)

Daywhite83 opened the pull request:
Gets the portals working, the catch is that all portals are open by default, this will need some input from others to fix. Once this one gets approved i can upload the others. (have them all made for all 3 zones)

<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?
Gets the portals working, the catch is that all portals are open by default, this will need some input from others to fix.
Once this one gets approved i can upload the others. (have them all made for all 3 zones)

I added some comments into the file explaining some of the things i have observed, which i can remove if needed.

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes

Go to Escha - Ru'Aun and use Eschan Portal No.1. (make sure you have at least 50 silt)

<!-- Clear and detailed steps to test your changes here -->

Daywhite83 committed:
DRAFT: Eschan_Portal_#1 (Escha-Ru'Aun)

Gets the portals working, the catch is that all portals are open by default, this will need some input from others to fix.
Once this one gets approved i can upload the others. (have them all made for all 3 zones)

zach2good commented on scripts/zones/Escha_RuAun/npcs/Eschan_Portal_#1:
```diff
@@ -0,0 +1,31 @@
+end
+
+entity.onEventFinish = function(player, csid, option)
+    player:delCurrency("escha_silt", siltCost)
```
If you bail out of the event, this cost will still be taken. You need to check against option

Daywhite83 committed:
fixed silt issue

added if statement for option to take silt only if option is 1

cocosolos commented:
It looks like the portal array is a bit field where bits 0-7 are Zi'Tah, 8-22 are Ru'Aun, and the rest are Reisenjima. e.g. 11111101 = 253 = Zi'Tah all portals unlocked except portal 2. Probably needs a new column in char_unlocks and everything that comes with that. Silt cost looks like its effected by zone blessings from domain invasion.

cocosolos commented:
Also it looks like option 5 after `currentPortalIndex` is a bool to trigger the unlock message (Light begins to shimmer and twist as you pass your hand over it.)

Daywhite83 committed:
fixed menu

played around with it a little bit and referred back to my retail captures a bit more closely. The menu's should work like retails now.

Daywhite83 commented:
option 5 seems to be the option for eschan droplets 3 = have it, 0 = no, 1 seem to have no effect here or it also means no. 2 also seems like a yes, but from captured data it used bit 3 so thats what i did as well

the portals feel retail accurate on my end now, fixed being able to see all portals by default, and they seem to unlock as you open them by retail methods. here are the portals 2-15 if u want to test for yourself.
[EschanPortals2-15.zip](https://github.com/LandSandBoat/server/files/10044838/EschanPortals2-15.zip)

cocosolos commented:
All 3 Escha zones share the same portal unlock variable. It is an [array of bits](https://en.wikipedia.org/wiki/Bit_array) where each bit gets flipped as you unlock the corresponding portal. Option 5 is also comprised of bits. If none of the bits are set then the value is 0 and you get the normal menu. If the first bit is set the value is 1 and you get the portal unlock message. If the first bit is not set and the second bit is, the value is 2 and you get the eschan droplets option without the unlock message. If both the first and second bits are set the value is 3 and you get the unlock message and the eschan droplets option.
![image](https://user-images.githubusercontent.com/2593549/202809277-97eb8ece-8458-4488-a57d-6232d184d166.png)

To implement this properly you'll probably need to make use of some of the Lua bitwise operator functions like `bit.bor`, `bit.band`, `bit.lshift`, `bit.rshift`, etc.

The eschan droplets option should only be enabled if the player _has_ eschan droplets from the mission, and care needs to be taken to give and take them properly as the player progresses.

The portals should be handled the same way homepoints, survival guides, abyssea confluxes, waypoints and other unlockables are handled in the char_unlocks table of the database and using the `addTeleport` and `getTeleport` functions, not with a charvar. This will require a little C++ work to add escha portal fields as well as adding it to the SQL file and a Python migration.

You'll also want to check and follow the [project Lua style guide](https://github.com/LandSandBoat/server/wiki/Development-Guide-Lua).

Xaver-DaRed commented:
This seems abandoned. It's also obsolete by now.
